README.md
OpenData Portal
Finite Software Systems

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Added - for new features.
Changed - for changes in existing functionality.
Deprecated - for soon-to-be removed features.
Removed - for now removed features.
Fixed - for any bug fixes.
Security - in case of vulnerabilities.

## [Unreleased] - XXXX-XX-XX

## [1.4.7] - 2021-11-18
### Fixed
- Publish dataset with recourses of type zip
- Docker tool to work with testdata.egov.bg

## [1.4.6] - 2021-11-05
### Added
- Docker tool to work with testdata.egov.bg

## [1.4.5] - 2021-11-05
### Added
- Matomo tracking code

### Changed
- api listDatasets default filters to match site Data section
- Info logs to use different file info.log instead of laravel.log

## [1.4.4] - 2021-10-05
### Changed
- Dataset create form visibility defaults to public

## [1.4.3] - 2021-10-05
### Added
- Dataset list access filter - free and paid

### Removed
- Dataset list filters by user and groups

## [1.4.2] - 2021-08-20
### Added
- Upload resource in zip format to data sets

### Fixed
- Zip download formats in dataset view

## [1.4.1] - 2021-08-18
### Added
- Send a notification emails to admins when a resource report signal is send
- Send a notification emails to admins when a data request is send

### Fixed
- Emails list in warning email when Elasticsearch cluster's node is down

## [1.4.0] - 2021-07-23
### Added
- Artisan command to check if all the Elasticsearch cluster's nodes are in good health 
  and if a node is down send warning emails to admins
- Artisan command for reindexing resources in Elasticsearch
- Move all resources from one data set to another and reindex the resources in Elasticsearch

### Changed
- Elasticsearch mapping from dynamic true to false when creating new resource and when 
  reindexing (prevent issue of mapping explosion) 

### Fixed
- News type in add and edit news. On delete check if news type is for latest news 
  and if so clear the cache for the latest news on the home page
- Last updated date on the dateset page, it doesn't get updated when updating resource via API

## [1.3.0] - 2021-06-28
### Added
- News type to news that can be regular, latest news for home page and alert
- Latest news box to the home page
- Alert news displayed just below the horizontal menu in the colored lane
- Resources direct download links for the different formats to the resource's details page
- Error pages 404, 500, 505
- Download links for zip files in a different format with all the resources in a data set

### Changed
- Replace registered user's count in the green box on the home page with organizations count
- Disable the first three fields in organization's edit form for all users except admins
- Temporarily remove the restriction of maximum 800 characters for news body
- Make an excerpt from the first 3000 characters of the resource preview when it is in json
  or xml format on the resource's details page

### Fixed
- Typo in data sets text on home page

### Removed
- Home page boxes for organizations count and updates count

## [1.2.5] - 2020-09-18
### Changed
- Report Elasticsearch red indices command

## [1.2.4] - 2020-09-17
### Changed
- Report Elasticsearch red indices command

## [1.2.3] - 2020-09-16
### Fixed
- Tool installation files

## [1.2.2] - 2020-09-14
### Added
- Contact phone to users
- reCaptcha to data requests and log in forms
- Data set setting for publishing data without admin approval

### Fixed
- Social media buttons on mobile screen
- Resource pagination info

### Changed
- Маке data set uri readonly field with auto generated value
- Disable user registration functionality

## [1.2.1] - 2020-08-31
### Added
- Report Elasticsearch red indices command

## [1.2] - 2020-08-20
### Fixed
- User rights issue

### Added
- COVID-19 resource visualizations
- Dynamic addition of sections in the footer area
- Name and information about the Portal in the header area
- Send information to Graylog for resource data requests through the API
- Merged pull request #28
- reCAPTCHA v2 checkbox

### Changed
- Social media buttons functionality
- Make "Organisation" mandatory field when publishing data

## [1.1.22] - 2020-03-12
### Fixed
- Removal of user rights

## [1.1.21] - 2020-03-04
### Fixed
- Incorrect validation to edit user

## [1.1.20] - 2020-02-28
### Fixed
- A few typos
- Multi-user registrations

### Changed
- Ordering of public documents

## [1.1.19] - 2019-12-19
### Changed
- Updated_by property controlled through trait

## [1.1.18] - 2019-12-19
### Fixed
- Filename error during download of resource

## [1.1.17] - 2019-11-19
### Removed
- Temporarily remove signal button

## [1.1.16] - 2019-11-14
### Removed
- Temporarily disable signal creation

## [1.1.15] - 2019-10-01
### Added
- Commands for elasticsearch data backup

### Fixed
- Relation retrieval for action history
- User e-mail invite button

## [1.1.14] - 2019-09-12
### Added
- Conversion for html files

### Fixed
- Editing datasets through API without org_id
- Newsletter object error

## [1.1.13] - 2019-09-10
### Added
- Option to manipulate xml size limits and field depth

### Changed
- Timeout for dataset publish

## [1.1.12] - 2019-09-03
### Added
- Option to manipulate resource mapping

## [1.1.11] - 2019-08-20
### Added
- Setup of multiple elastic hosts
- Additional validation for dataset publish

### Fixed
- Subcategory deletion redirect
- Public filter interactions and ordering
- Load times of history of actions, logins by setting default dates
- Query for most active organisation
- Limit of organisation list for data requests

## [1.1.10] - 2019-07-23
### Added
- Command for removal of elastic indices without metadata
- Command for removal of elastic indices deleted during migration

### Fixed
- Load speed of notifications page

## [1.1.9] - 2019-07-17
### Fixed

app/Module.php
<?php

namespace App;

use App\ActionsHistory;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Auth;
use Illuminate\Database\Eloquent\Model;


class Module extends Model
{
    /*
     * User actions
     */
    const ORGANISATIONS = 1;
    const GROUPS = 2;
    const USERS = 3;
    const DATA_SETS = 4;
    const RESOURCES = 5;
    const SIGNALS = 6;
    const DATA_REQUESTS = 7;
    const MAIN_CATEGORIES = 8;
    const ACTIONSHISTORY = 9;
    const SECTIONS = 10;
    const NEWS = 11;
    const ROLES = 12;
    const PAGES = 13;
    const DOCUMENTS = 14;
    const TERMS_OF_USE = 15;
    const TERMS_OF_USE_REQUESTS = 16;
    const LOCALE = 17;
    const DATA_CONVERSIONS = 18;
    const HELP = 19;
    const IMAGES = 20;
    const TAGS = 21;
    const HELP_SECTIONS = 22;
    const HELP_PAGES = 23;
    const CUSTOM_SETTINGS = 24;
    const RIGHTS = 25;
    const THEMES = 26;
    const TOOL_DB_CONNECTION = 27;
    const TOOL_FILE = 28;
    const TOOL_DB_QUERY = 29;
    const MODULES = 30;

    /**
     * Gives back the available user actions.
     *
     * @return Array with user actions as key-value pairs
     */
    public static function getModules()
    {
        return [
            self::ORGANISATIONS          => 'Organisation',
            self::GROUPS                 => 'Group',
            self::USERS                  => 'User',
            self::DATA_SETS              => 'Dataset',
            self::RESOURCES              => 'Resource',
            self::SIGNALS                => 'Signal',
            self::DATA_REQUESTS          => 'DataRequest',
            self::MAIN_CATEGORIES        => 'MainCategories',
            self::ACTIONSHISTORY         => 'ActionsHistory',
            self::TAGS                   => 'Tags',
            self::SECTIONS               => 'Section',
            self::NEWS                   => 'News',
            self::ROLES                  => 'Role',
            self::PAGES                  => 'Page',
            self::DOCUMENTS              => 'Document',
            self::TERMS_OF_USE           => 'TermsOfUse',
            self::TERMS_OF_USE_REQUESTS  => 'TermsofUseRequests',
            self::LOCALE                 => 'Locale',
            self::DATA_CONVERSIONS       => 'DataConversion',
            self::HELP_SECTIONS          => 'HelpSection',
            self::HELP_PAGES             => 'HelpPage',
            self::IMAGES                 => 'Image',
            self::CUSTOM_SETTINGS        => 'CustomSetting',
            self::RIGHTS                 => 'Right',
            self::THEMES                 => 'Theme',
            self::MODULES                => 'Modules',
        ];
    }

    public static function getToolModules()
    {
        return [
            self::TOOL_DB_CONNECTION    => 'DB Connection',
            self::TOOL_DB_QUERY         => 'DB Query',
            self::TOOL_FILE             => 'File',
        ];
    }

    public static function getModuleName($moduleIndex)
    {
        $modules = config('app.IS_TOOL') ? self::getToolModules() : self::getModules();

        if (in_array($moduleIndex, array_flip($modules))) {
            return $modules[$moduleIndex];
        }

        return false;
    }

    /**
     * Record action history by module and action for a logged user
     *
     * @param string moduleName - comming from MODULE_NAMES (required)
     * @param integer type - comming from TYPE_ constants (required)
     * @param string|integer object - comming from the action object constants or is custom string (required)
     * @param string message - used to describe the taken action (required)
     *
     * @return boolean wheather user is authorized or not
     */
    public static function add($request)
    {
        if (config('app.IS_TOOL') || Auth::check()) {
            $actions = ActionsHistory::getTypes();

            $validator = \Validator::make($request, [
                'module_name'   => 'required|string|max:191',
                'action'        => 'required|int|digits_between:1,3|in:'. implode(',', array_flip($actions)),
                'action_msg'    => 'required|string|max:191',
            ]);

            $actionObject = isset($request['action_object']) ? $request['action_object'] : '';
            $ip = request()->ip();

            if (!$validator->fails()) {
                try {
                    $dbData = [
                        'module_name'   => $request['module_name'],
                        'action'        => $request['action'],
                        'action_object' => $actionObject,
                        'action_msg'    => $request['action_msg'],
                        'ip_address'    => !empty($ip) ? $ip : 'N/A',
                        'user_agent'    => isset($_SERVER['HTTP_USER_AGENT']) ? $_SERVER['HTTP_USER_AGENT'] : 'N/A',
                        'occurrence'    => date('Y-m-d H:i:s'),
                    ];

                    if (!config('app.IS_TOOL') && !config('app.IS_TEST_TOOL')) {
                        $dbData['user_id'] = Auth::user()->id;
                    }

                    if (isset($request['status'])) {
                        $dbData['status'] = $request['status'];
                    }

                    ActionsHistory::create($dbData);

                } catch (QueryException $ex) {
                    Log::error($ex->getMessage());
                }
            }
        }
    }
}
